### PR TITLE
Fix QuantizeLinear shape inference

### PR DESCRIPTION
**Pull request by philass**
Fix QuantizeLinear shape inference in cases where type needs to be read from zero point attribute

### CODE REVIEW

File: src/Dialect/ONNX/ONNXOps.cpp [hunk 1]
```diff
@@ -3267,10 +3267,18 @@ LogicalResult ONNXQuantizeLinearOp::inferShapes(
...
-    // TODO: Unfortunately, we can't tell if this should be signed or
-    // unsigned here...
-    IntegerType i8Type = IntegerType::get(getContext(), 8);
-    RankedTensorType outType = RankedTensorType::get(inTy.getShape(), i8Type);
+    auto zero = y_zero_point();
+    RankedTensorType outType;
+    // If zero point type isn't provided, output type defaults to ui8.
+    if (isFromNone(zero)) {
+      auto builder = mlir::Builder(this->getContext());
+      IntegerType ui8Type = builder.getIntegerType(8, false);
```
Comment (sorenlassen) [inline src/Dialect/ONNX/ONNXOps.cpp:3275]: nit: I think the most readable and unambiguous way to express this (and gets rid of the builder) is:
```
IntegerType::get(getContext(), 8, IntegerType::Unsigned)
```

File: src/Dialect/ONNX/ONNXOps.cpp [hunk 2]
```diff
@@ -3267,10 +3267,18 @@ LogicalResult ONNXQuantizeLinearOp::inferShapes(
...
+      outType = RankedTensorType::get(inTy.getShape(), ui8Type);
+    } else {
+      // If zero point is provided, output type is same as zero point type.
+      auto zpTy = zero.getType().cast<ShapedType>().getElementType();
+      outType = RankedTensorType::get(inTy.getShape(), zpTy);
```
Comment (sorenlassen) [inline src/Dialect/ONNX/ONNXOps.cpp:3280]: nit: to make naming consistent `zpTy` should refer to its tensor type, so you should write:
```
auto zpTy = zero.getType().cast<ShapedType>();
outType = RankedTensorType::get(inTy.getShape(), zpTy.getElementType());
```

File: test/mlir/onnx/onnx_shape_inference.mlir [hunk 3]
```diff
@@ -1555,6 +1555,16 @@ func.func @test_quantize_linear_1(%arg0 : tensor<5x2x3x4xf32>, %arg1 : tensor<f3
   // CHECK: return [[RES]] : tensor<5x2x3x4xi8>
 }
 
+func.func @test_quantize_linear_2(%arg0 : tensor<5x2x3x4xf32>, %arg1: tensor<f32>, %arg2: tensor<ui8>) -> tensor<*xui8> {
```
Comment (sorenlassen) [inline test/mlir/onnx/onnx_shape_inference.mlir:1558]: for completeness you could add one more lit test with y_zero_point `None`

**Review by sorenlassen — APPROVED** [review]

File: src/Dialect/ONNX/ONNXOps.cpp (hunk 1 above)
Comment (philass) [reply src/Dialect/ONNX/ONNXOps.cpp:3275]: Thanks! I was struggling to find `IntegerType` in the mlir docs. I saw the function on the builder.

But your suggestion is definitely better.

File: src/Dialect/ONNX/ONNXOps.cpp [hunk 4]
```diff
@@ -3267,10 +3267,18 @@ LogicalResult ONNXQuantizeLinearOp::inferShapes(
...
+    // If zero point type isn't provided, output type defaults to ui8.
+    if (isFromNone(zero)) {
+      IntegerType ui8Type =
+          IntegerType::get(getContext(), 8, IntegerType::Unsigned);
+      outType = RankedTensorType::get(inTy.getShape(), ui8Type);
+    } else {
+      // If zero point is provided, output type is same as zero point type.
+      auto zpTy = zero.getType().cast<ShapedType>();
+      outType = RankedTensorType::get(inTy.getShape(), zpTy.getElementType());
+    }
```
Comment (tungld) [inline src/Dialect/ONNX/ONNXOps.cpp:3281]: Just wanted to make this more concise, how about 
```C
Type elementType;
if (isFromNone(zero))
  elementType = IntegerType::get(getContext(), 8, IntegerType::Unsigned);
else
  elementType = zero.getType().cast<ShapedType>().getElementType();
auto outType = RankedTensorType::get(inTy.getShape(), elementType);
```

File: src/Dialect/ONNX/ONNXOps.cpp [hunk 5]
```diff
@@ -3267,10 +3267,18 @@ LogicalResult ONNXQuantizeLinearOp::inferShapes(
   auto yTy = y().getType().cast<ShapedType>();
 
   if (!yTy.hasStaticShape()) {
-    // TODO: Unfortunately, we can't tell if this should be signed or
-    // unsigned here...
-    IntegerType i8Type = IntegerType::get(getContext(), 8);
-    RankedTensorType outType = RankedTensorType::get(inTy.getShape(), i8Type);
+    auto zero = y_zero_point();
```
Comment (tungld) [inline src/Dialect/ONNX/ONNXOps.cpp:3270]: We are trying to avoid `auto`, it looks like its type is `Value` here.

File: src/Dialect/ONNX/ONNXOps.cpp (hunk 4 above)
Comment (philass) [reply src/Dialect/ONNX/ONNXOps.cpp]: Thanks for the suggestion. That looks a lot cleaner
